feat: upgrade function extension and extension bundle

pudkrong opened the pull request:
Azure functions runtime version 3 reached end of life on 22 Dec 2022. So upgrade to latest version, 4 and node version to 18

pudkrong committed:
feat: upgrade function extension and extension bundle

Azure functions runtime version 3 reached end of life on 22 Dec 2022.
So upgrade to latest version, 4 and node version to 18

fix: revert @types/node

coderbyheart commented on feature-runner/steps/device/connectDevice.ts:
```diff
@@ -29,20 +29,34 @@ export const connectDevice = async ({
+		deviceKey,
+		deviceCert,
+		intermediateCA,
+		digiCert,
+		baltimore,
+		microsoftRSARoot,
```
Since we are not migrating devices, we should only need the new DigiCertGlobalRootG2.pem, not the other two.

pudkrong replied:
> Since we are not migrating devices, we should only need the new DigiCertGlobalRootG2.pem, not the other two.

According to MS docs, they said `The Microsoft RSA Root Certificate Authority 2017 helps prevent future disruptions in case the DigiCert Global Root G2 is retired unexpectedly.`

What do you think?

coderbyheart replied:
That's not relevant for the software-based device used in the end-to-end test runner.

pudkrong committed:
fix: jsonwebtoken fail to sign using ES256

jsonwebtoken lib requires curve 'prime256v1' when using 'ES256' algorithm,
auth0/node-jsonwebtoken#862

pudkrong committed:
feat: update TLS certificates

Reference: https://learn.microsoft.com/en-us/azure/iot-hub/migrate-tls-certificate?tabs=portal

fix: iot hub still requires baltimore certificate

coderbyheart commented on .github/workflows/test-and-release.yaml:
```diff
@@ -30,6 +30,7 @@ jobs:
       APP_REG_CLIENT_ID: ${{ secrets.APP_REG_CLIENT_ID }}
       B2C_CLIENT_SECRET: ${{ secrets.B2C_CLIENT_SECRET }}
       B2C_TENANT_ID: ${{ secrets.B2C_TENANT_ID }}
+      B2C_USER_FLOW_NAME: B2C_1_developer
```
I assume this change is needed for CI to work.

pudkrong replied:
Previously, it is hardcode. So, making it as parameter will give us more flexible to test somewhere else

pudkrong replied:
This change is needed for CI to work

coderbyheart commented on .github/workflows/test-and-release.yaml:
```diff
@@ -110,11 +111,11 @@ jobs:
           inlineScript: |
             az keyvault secret set --vault-name ${keyVaultName} --name unwiredlabsApiKey --value my-secret
-            openssl ecparam -name secp256k1 -genkey > nrfcloudcelllocationservice.key
+            openssl ecparam -name prime256v1 -genkey > nrfcloudcelllocationservice.key
```
Unrelated change.

pudkrong replied:
It is required to fix the error on https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/e0bc28ebb5a4d0dd04d2d5e66c8610c6d9f13324/third-party/nrfcloud.com/apiclient.ts#L422. 

According to jsonwebtoken lib requires curve 'prime256v1' when using 'ES256' algorithm,
https://github.com/auth0/node-jsonwebtoken/issues/862

coderbyheart replied:
So, it's unrelated but required for CI to pass?

pudkrong replied:
Yes

coderbyheart commented on data/BaltimoreCyberTrustRoot.pem:
```diff
@@ -18,4 +18,4 @@ jkzSswF07r51XgdIGn9w/xZchMB5hbgF/X++ZRGjD8ACtPhSNzkE1akxehi/oCr0
 R9I4LtD+gdwyah617jzV/OeBHRnDJELqYzmp
------END CERTIFICATE-----
\ No newline at end of file
+-----END CERTIFICATE-----
```
Unrelated change.

coderbyheart commented on data/DigiCertGlobalRootG2.pem:
I assume this is needed for CI to work.

pudkrong replied:
yes, it is required to enable device simulator to connect to IoT hub

coderbyheart commented on geolocateCellFromNrfCloud/geolocateCell.ts:
```diff
@@ -142,6 +143,7 @@ const geolocateCell: AzureFunction = async (
 					404,
 				)
 				context.bindings.cellGeolocation = JSON.stringify({
+					id: randomUUID(),
```
Use cellID as id.

coderbyheart commented on mock-http-api.json:
```diff
@@ -75,15 +75,15 @@
             {
               "name": "WEBSITE_NODE_DEFAULT_VERSION",
-              "value": "~14"
+              "value": "~18"
```
Can we updated to Node 20?

pudkrong replied:
It is still in public preview, https://azure.microsoft.com/en-us/updates/public-preview-azure-functions-support-for-nodejs-20/#:~:text=Published%20date%3A%20September%2027%2C%202023&text=You%20can%20now%20develop%20functions,your%20Function%20Apps%20to%20Node. So, do you still want to use node20?

pudkrong replied:
Another thing is, if we use node20, there is warning error from `@azure/msal-node@1.24.2` as

```
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@azure/msal-node@1.14.2',
npm WARN EBADENGINE   required: { node: '10 || 12 || 14 || 16 || 18' },
npm WARN EBADENGINE   current: { node: 'v20.8.0', npm: '9.7.2' }
npm WARN EBADENGINE }
```

coderbyheart replied:
Yes, because it is LTS in 12 days: https://github.com/nodejs/release#release-schedule

pudkrong replied:
To upgrade to Node v20, Azure function has to migrate to programming model v4 which has a big change. So, we will use Node v18 for now. It will be another PR to upgrade to Node v20.

coderbyheart commented on package.json:
```diff
@@ -59,7 +59,7 @@
     "@types/jest": "29.4.0",
     "@types/jsonwebtoken": "9.0.1",
-    "@types/node": "18.14.0",
+    "@types/node": "18.18.0",
```
Unrelated change.

pudkrong replied:
Using version `18.14.0`, I have found error about type mismatch on `node-fetch` at https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/e0bc28ebb5a4d0dd04d2d5e66c8610c6d9f13324/feature-runner/lib/b2c/accessToken.ts#L58

The error message is

```
Type 'URLSearchParams' is not assignable to type 'BodyInit'.
  Property 'size' is missing in type 'import("url").URLSearchParams' but required in type 'URLSearchParams'.ts(2322)
lib.dom.d.ts(22575, 14): 'size' is declared here.
index.d.ts(75, 2): The expected type comes from property 'body' which is declared here on type 'RequestInit'
```

coderbyheart replied:
Get rid of node-fetch and use native fetch (which is available since v18)

pudkrong replied:
After removing `node-fetch` and use native fetch, I can then revert `@types/node`changes.

coderbyheart commented on storeDeviceCellGeolocation/storeDeviceCellGeolocation.ts:
Remove device cell geo location feature, see https://github.com/NordicSemiconductor/asset-tracker-cloud-aws-js/commit/1e0735453af47a2ebae0e6fecc88935e16244a7b

pudkrong committed:
feat: using cell id as id

pudkrong committed:
feat: using native fetch

pudkrong committed:
feat: remove cell geo location feature

coderbyheart commented on geolocateCellFromNrfCloud/geolocateCell.ts:
```diff
@@ -142,6 +142,7 @@ const geolocateCell: AzureFunction = async (
 					404,
 				)
 				context.bindings.cellGeolocation = JSON.stringify({
+					id,
 					cellId: id,
```
do not duplicate the value.

coderbyheart commented:
Great!

pudkrong committed:
Merge branch 'saga' into upgrade-azure-functions-runtime
